src/transaction/log_replication.cpp

```diff
@@ -204,6 +229,32 @@ namespace cublog
 	    m_undo_unzip, m_redo_unzip, m_parallel_replication_redo, true);
   }
 
+  template <typename T>
+  void replicator::calculate_replication_delay_or_dispatch_async (cubthread::entry &thread_entry,
+      const log_lsa &rec_lsa, const char *a_source)
+  {
+    const T log_rec = m_reader.reinterpret_copy_and_add_align<T> ();
+    // at_time, expressed in milliseconds rather than seconds
+    const time_msec_t start_time_msec = log_rec.at_time;
+    if (m_parallel_replication_redo != nullptr)
+      {
+	// dispatch a job; the time difference will be calculated when the job is actually
+	// picked up for completion by a task; this will give an accurate estimate of the actual
+	// delay between log genearation on the page server and log recovery on the page server
+	using redo_job_replication_delay_impl_t = cublog::redo_job_replication_delay_impl<T>;
+	std::unique_ptr<redo_job_replication_delay_impl_t> replication_delay_job
+	{
+	  new redo_job_replication_delay_impl_t (m_redo_lsa, start_time_msec, a_source)
+	};
+	m_parallel_replication_redo->add (std::move (replication_delay_job));
+      }
+    else
+      {
+	// calculate the time difference synchronously
+	log_rpl_calculate_replication_delay (&thread_entry, start_time_msec, a_source);

```

Comment:
Even in this case, shouldn't LOG_COMMIT and LOG_ABORT log records are processed?

Reply:
I do not think I understant your question. Can you please elaborate?
Some explanations though:
The else statement performs the replication delay calculation & perfstat reporting when the replication is working in the synchronous mode (ie: when the param `PRM_ID_REPLICATION_PARALLEL_COUNT` is set to `0`). We decided to leave this infrastructure able to also work - with full functionality - in synchronous mode as well.

Reply:
@hornetmj
After 3 days I finally undestand what you asked. They were not processed before I added handling to calculate the delay, so I suppose they do not need to be. I'm not knowledgeable in this matter but I suppose the following happens:
- the transactions are handled on the active transaction server and also introduce other log entries that are actually affect data which are properly processed;
- the COMMIT and ABORT log entries are only introduced such that certain bookkeeping is also performed (by analogy with the handling of `LOG_COMMIT` `LOG_ABORT` in log_recovery - `log_recovery_redo` function) and this bookkeeping is not needed on the page server;

@ribram
Please confirm this.

Reply:
Short answer, I confirm what @cristiarg says.

Long answer, as of now, the page server only replicated changes with effects on disk. The transaction table is not updated. Recovery also tracks transaction table changes, so it has some extra operations.